tcms/telemetry/api.py

```diff
@@ -169,3 +169,46 @@ def _append_status_counts_to_result(count, result):
     for status in TestExecutionStatus.chart_status_names:
         status_count = count.get(status, 0)
         result.get(status).append(status_count)
+
+
+@rpc_method(name='Testing.test_case_health')
+def test_case_health(query=None):
+
+    if query is None:
+        query = {}
+
+    all_test_executions = TestExecution.objects.filter(**query)
+
+    test_executions = _get_count_for(all_test_executions)
+    passed_test_executions = _get_count_for(all_test_executions,
+                                            TestExecutionStatus.passed_status_names)
+    failed_test_executions = _get_count_for(all_test_executions,
+                                            TestExecutionStatus.failure_status_names)
+
+    data = {}
+    for te in test_executions:
+        data[te['case_id']] = {
+            'case_id': te['case_id'],
+            'case_summary': te['case__summary'],
+            'count': {
+                'all': te['count'],
+                'success': 0,
+                'fail': 0
+            }
+        }
+
+    for te in passed_test_executions:
+        data[te['case_id']]['count']['success'] = te['count']
+    for te in failed_test_executions:
+        data[te['case_id']]['count']['fail'] = te['count']
```

Comment:
Somewhere in these for loop you will have to implement a bit of filtering. On a large installation the dataset returned here may be overwhelming.

- Everything with a 100% PASS rate is not interesting for telemetry purposes
- Everything with 0% PASS rate is interesting, but how do we sort/limit if there are more than 10-20 ? Maybe the first batch regardless of ordering is fine (ordering is by case_id anyway). If there are more than LIMIT they will show up after the current ones have been improved (We should probably start documenting that).
- Anything where pass rate is > 0 and < 100% is interesting. Not sure if we need the ones where PASS rate is closer to 100 (e.g. were mostly fine, but kind of failed a bit) or where pass rate is closer to 0 (were mostly failing with occasional pass here and there - this is definitely more actionable if you want a stable test suite). Or do both ?!?

Reply:
> Everything with a 100% PASS rate is not interesting for telemetry purposes

```python
for key in data.copy():
    if data[key]['count']['success'] == data[key]['count']['all']:
        data.pop(key)
```
this removes all the entries with 100% pass rate.

then we sort them so that the ones with the smallest passing rates are first:
```python
data.sort(key=_sort_by_passing_rate)

def _sort_by_passing_rate(element):
    return element['count']['success'] / element['count']['all']
```

then we collect the first 20 and return them:
```python
result = []
for i in range(0, 20):
    result.append(data[i])

return result
```

this way we get 20 entries sorted by passing rate ascending.

does this look ok to you? it added a few more list iterations in the code, but I don't think we can achieve that with SQL only